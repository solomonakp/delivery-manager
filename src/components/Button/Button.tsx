import Loader from 'components/Loader';
import React, { FC } from 'react';

type color = 'primary' | 'warning';

export interface ButtonProps
  extends React.DetailedHTMLProps<
      React.ButtonHTMLAttributes<HTMLButtonElement>,
      HTMLButtonElement
    >,
    React.AriaAttributes {
  loading?: boolean;
}

const Button: FC<ButtonProps> = (props) => {
  const { children, className, loading = false, ...rest } = props;

  return (
    <button className={`${className} btn`} {...rest}>
      {children}
      {loading && <Loader type={2} />}
    </button>
  );
};

export default Button;
