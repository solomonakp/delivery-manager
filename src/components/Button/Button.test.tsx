import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { axe } from 'jest-axe';
import '@testing-library/jest-dom/extend-expect';
import Button from './Button';
import { ButtonProps } from './Button';

const defaultProps: ButtonProps = {
  onClick: jest.fn(),
  title: 'my button',
  role: 'button',
  type: 'button',
};

afterEach(cleanup);

const setUpButton = (props: ButtonProps = defaultProps) =>
  render(<Button {...props}>click</Button>);

describe('<Button />', () => {
  test('should mount', () => {
    setUpButton();

    const button = screen.getByText(/click/i);

    expect(button).toBeInTheDocument();
  });

  test('should render the button user would see', () => {
    const { asFragment } = setUpButton();

    expect(asFragment()).toMatchSnapshot();
  });

  test('should fire event when clicked', () => {
    const { onClick } = defaultProps;

    setUpButton();

    const button = screen.getByText(/click/i);

    fireEvent.click(button);

    expect(onClick).toHaveBeenCalled();
  });

  test('shows a loader while loading', async () => {
    const props = {
      ...defaultProps,
      loading: true,
    };
    setUpButton(props);

    expect(await screen.findByTestId(/loader/i)).toBeInTheDocument();
  });
  test.skip('should not fail any accessibility test', async () => {
    const { container } = setUpButton();

    const result = await axe(container);

    expect(result).toHaveNoViolations();
  });
});
