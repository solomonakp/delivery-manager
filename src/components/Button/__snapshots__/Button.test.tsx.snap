// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button user would see 1`] = `
<DocumentFragment>
  <button
    class="undefined btn"
    role="button"
    title="my button"
    type="button"
  >
    click
  </button>
</DocumentFragment>
`;
